core/shared/src/main/scala/zio/ZEnvironment.scala

```diff
@@ -221,6 +229,84 @@ object ZEnvironment {
       System.SystemLive
     )
 
+  /**
+   * A `Patch[In, Out]` describes an update that transforms a `ZEnvironment[In]`
+   * to a `ZEnvironment[Out]` as a data structure. This allows combining updates
+   * to different services in the environment in a compositional way.
+   */
+  sealed trait Patch[-In, +Out] { self =>
+    import Patch._
+
+    /**
+     * Applies an update to the environment to produce a new environment.
+     */
+    def apply(environment: ZEnvironment[In]): ZEnvironment[Out] = {
+
+      @tailrec
+      def loop(environment: ZEnvironment[Any], patches: List[Patch[Any, Any]]): ZEnvironment[Any] =
+        patches match {
+          case AddService(service, tag) :: patches   => loop(environment.unsafeAdd(tag, service), patches)
+          case AndThen(first, second) :: patches     => loop(environment, erase(first) :: erase(second) :: patches)
+          case Empty() :: patches                    => loop(environment, patches)
+          case RemoveService(tag) :: patches         => loop(environment, patches)
+          case UpdateService(update, tag) :: patches => loop(environment.unsafeUpdate(tag, update), patches)
+          case Nil                                   => environment
+        }
+
+      loop(environment, List(self.asInstanceOf[Patch[Any, Any]])).asInstanceOf[ZEnvironment[Out]]
+    }
+
+    /**
+     * Combines two patches to produce a new patch that describes applying the
+     * updates from this patch and then the updates from the specified patch.
+     */
+    def combine[Out2](that: Patch[Out, Out2]): Patch[In, Out2] =
+      AndThen(self, that)
+  }
+
+  object Patch {
+
+    /**
+     * An empty patch which returns the environment unchanged.
+     */
+    def empty[A]: Patch[A, A] =
+      Empty()
+
+    /**
+     * Constructs a patch that describes the updates necessary to transform the
+     * specified old environment into the specified new environment.
+     */
+    def fromDiff[In, Out](oldValue: ZEnvironment[In], newValue: ZEnvironment[Out]): Patch[In, Out] = {
+      val sorted = newValue.map.toList.sortBy { case (_, (_, index)) => index }
+      val (missingServices, patch) = sorted.foldLeft[(Map[LightTypeTag, Any], Patch[In, Out])](
+        oldValue.map -> Patch.Empty().asInstanceOf[Patch[In, Out]]
+      ) { case ((map, patch), (tag, (newService, _))) =>
+        map.get(tag) match {
```

Comment:
If possible, we should have a couple tests to ensure that patching ZEnvironment is well-defined.